src/compiler/transformers/es2015.ts

```diff
@@ -2610,7 +2620,40 @@ namespace ts {
             return visitEachChild(node, visitor, context);
         }
 
-        function shouldConvertIterationStatementBody(node: IterationStatement): boolean {
+        interface ForStatementWithConvertibleInitializer extends ForStatement {
+            initializer: VariableDeclarationList;
+        }
+
+        interface ForStatementWithConvertibleCondition extends ForStatement {
+            condition: Expression;
+        }
+
+        interface ForStatementWithConvertibleIncrementor extends ForStatement {
+            incrementor: Expression;
+        }
+
+        function shouldConvertPartOfIterationStatement(node: Node) {
+            return (resolver.getNodeCheckFlags(node) & NodeCheckFlags.ContainsCapturedBlockScopeBinding) !== 0;

```

Comment:
I dislike a non-TS transform relying on typechecking information to be correct. Especially since this means that in the presence of a `before` transformer, it's very possible that we'll fail to convert something which needs to be.

Reply:
We do this in other places, which is the reason we have `original` pointers for nodes to begin with.

Reply:
Geh, you're right. It's actually only a few transforms that can operate without a typechecker available.